### PR TITLE
Support custom build commands or script for Nodejs apps

### PR DESCRIPTION
**Pull request by qianz2**
- [x] The purpose of this PR is explained in this message or in an issue. If an issue please include a reference as \#\<issue\_number\>.
- [x] Tests are included and/or updated for code changes.
- [x] Proper license headers are included in each file.

Issue #823

### CODE REVIEW

File: src/BuildScriptGenerator/Node/NodeScriptGeneratorOptions.cs [hunk 1]
```diff
@@ -13,6 +13,12 @@ public class NodeScriptGeneratorOptions
         /// </summary>
         public string CustomRunBuildCommand { get; set; }
 
+        /// <summary>
+        /// Gets or sets custom build command or script that will run without 'npm install' in the generated
+        /// build script.
+        /// </summary>
```
Comment (arroyc) [inline src/BuildScriptGenerator/Node/NodeScriptGeneratorOptions.cs:19]: I'm bit confused .. what is the difference between a customrunbuildcommand and a custombuildcommand

File: src/BuildScriptGenerator/Node/NodeBashBuildSnippet.sh.tpl [hunk 2]
```diff
@@ -93,7 +93,12 @@ fi
 
 cd "$SOURCE_DIR"
 
-{{ if CustomRunBuildCommand | IsNotBlank }}
+{{ if CustomBuildCommand | IsNotBlank }}
+	echo
+	echo "Running '{{ CustomBuildCommand }}'..."
+	echo
+	{{ CustomBuildCommand }}
+{{ else if CustomRunBuildCommand | IsNotBlank }}
```
Comment (arroyc) [inline src/BuildScriptGenerator/Node/NodeBashBuildSnippet.sh.tpl:101]: so you are prioritizing custombuildcommand over customrunbuildcommand .... Can you site an example/use case?

Comment (arroyc) [reply src/BuildScriptGenerator/Node/NodeBashBuildSnippet.sh.tpl:101]: I think we need to understand it better in terms of what we are looking to do here and what npm already offers .. may be we  can leverage npm to do this what we are trying to do here.
Let's discuss this over standup ...

**Review by arroyc — CHANGES_REQUESTED** [review]

**Comment by arroyc** [comment]
so let's say we have this in the package.json ... 

> "scripts": {
>   "lint": "jshint **",
>   "build:css": "stylus assets/styles/main.styl > dist/main.css",
>   "build:js": "browserify assets/scripts/main.js > dist/main.js",
>   "build": "npm run build:css && npm run build:js",
>   "prebuild:js": "npm run lint"
> }

https://docs.npmjs.com/cli/v6/using-npm/scripts

I want to see how the customrunbuild or custombuild command helps us compared to just running npm install what we are doing today.

File: src/BuildScriptGenerator/Node/NodeScriptGeneratorOptions.cs (hunk 1 above)
Comment (qianz2) [reply src/BuildScriptGenerator/Node/NodeScriptGeneratorOptions.cs:19]: So the difference is basically custombuildcommand will override any PackageInstallCommand Oryx provided(either could be 'npm install' or 'yarn install')

File: src/BuildScriptGenerator/Node/NodeBashBuildSnippet.sh.tpl (hunk 2 above)
Comment (qianz2) [reply src/BuildScriptGenerator/Node/NodeBashBuildSnippet.sh.tpl:101]: https://docs.npmjs.com/cli/v6/commands/npm-ci
Currently I have two examples, one is usage of npm ci command, which allows to install dependencies in automated environment(more strict and helps catch inconsistencies). The other one is an app that containing a `Makefile`, which has its own build command, like `make build`.

**Review by arroyc — APPROVED** [review]

**Pull request merged**